Enhancement - only save keys to settings.providers if validated

kliu57 opened the pull request:
Fixes #596 

**Summary:** No longer saving providers or their api keys to `settings.providers` array unless the api key is validated. If an invalid or blank is entered, the provider is REMOVED from `settings.providers`.

**Important coding points:**

- Removed the placeholder text in the api key field when you create a new custom provider row (placeholder text was showing up as asterix)
- Removed the `useDebounce` code, it wasn't working and is not worth the trouble to fix. `useDebounce` is sort of like a `useEffect`, but which allows us to wait until the user stops typing to validate the key. However, we do not expect anyone to be typing out a 32 character api key anyways, so opted to remove.
- The keys are stored in `tableProviders` and only when they are validated they are stored to `settings.providers`
- In these scenarios, `tableProviders` is synced with `settings.providers`: (when "syncing" happens, those invalid and blank api keys will disappear)
   - When modal first opens
   - When modal closes
   - When current provider is set (need to sync to update the checkmark)
   - When a custom provider is deleted (need to sync to show the row deleted)
   - When a custom provider is added (need to sync to show the new row)
- **Note:** If you create a custom provider and then later set the key to something invalid, when you close the modal the custom provider row will disappear since we only save it if it has a valid key.

I put in some `console.log` comments for @rjwignar ; those will be removed before I merge to main

kliu57 commented:
@rjwignar @Rachit1313 Sorry not yet ready for review. Will comment again when it is.

kliu57 commented:
@rjwignar @Rachit1313 Now it is ready for review

WangGithub0 commented:
I found the error color changed, is this the expected?
production:
<img width="530" alt="image" src="https://github.com/tarasglek/chatcraft.org/assets/20795443/f875dc0f-e0e4-4064-bba1-8695ad9ef072">

priview
<img width="616" alt="image" src="https://github.com/tarasglek/chatcraft.org/assets/20795443/d1a1a801-3619-46cf-989a-26386e5963f4">

kliu57 commented:
> I found the error color changed, is this the expected? production: <img alt="image" width="530" src="https://private-user-images.githubusercontent.com/20795443/323331981-f875dc0f-e0e4-4064-bba1-8695ad9ef072.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTMzNzc5OTksIm5iZiI6MTcxMzM3NzY5OSwicGF0aCI6Ii8yMDc5NTQ0My8zMjMzMzE5ODEtZjg3NWRjMGYtZTBlNC00MDY0LWJiYTEtODY5NWFkOWVmMDcyLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA0MTclMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNDE3VDE4MTQ1OVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTU1ZDkzMzQ3Mjg0MTU2MzI5MzFjZjZjNmQ2YmI4Mzk0ZDliYjM5YWNjZmNiNmUyOWYzMmQ1MTFlZTYwNWZhZjUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.8rYioRUAR4BX-vBDeKtOYEVDT9w8QooLncYaGm8LEgQ">
> 
> priview <img alt="image" width="616" src="https://private-user-images.githubusercontent.com/20795443/323331923-d1a1a801-3619-46cf-989a-26386e5963f4.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTMzNzc5OTksIm5iZiI6MTcxMzM3NzY5OSwicGF0aCI6Ii8yMDc5NTQ0My8zMjMzMzE5MjMtZDFhMWE4MDEtMzYxOS00NmNmLTk4OWEtMjYzODZlNTk2M2Y0LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA0MTclMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNDE3VDE4MTQ1OVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWNmOWFhNmY0YTE4MDNkZDRhMTI5OWQwZjliNjliYTcyYTNhODg2NTI5ZTBiMmVjNjk1ZjExZWRkOTZkZjc5MDYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.LPTltGPKGt0CjrioM5a_FW_TudDlODvwLqhMJMlSyAg">

The color is different because the message is different. The orange one happens when you try to add a new custom provider, but you put an invalid key. So it says "Provider not added" I used a warning instead of error so it can automatically disappear, I feel it is more convenient. The red one is when you change your current provider key to an invalid one.

Rachit1313 reviewed (approved):
Looks good to me.

tarasglek commented:
I think Error was more correct here..as it's fairly fatal that you can't validate your key

kliu57 commented:
> I think Error was more correct here..as it's fairly fatal that you can't validate your key

@tarasglek Ok, I modified the code so that users trying to add a new custom provider but gave an invalid url or invalid key will get the "error" type alert instead of the warning

kliu57 commented:
@WangGithub0 I changed the alert type when you create new custom providers with erroneous info to error instead of warning. 

Also I found an issue where the placeholder text of the `PasswordInput` was displaying as asterix and I fixed it

kliu57 commented:
@WangGithub0 also I found an issue where the placeholder text was showing up as asterix after I made the change in last hotfix PR to the `PasswordInput` component to stop browsers from asking users to save their password. Could not find an easy fix for this so for now I removed the "API Key" placeholder text for the password field.

WangGithub0 reviewed (approved):
It looks good to me.
